File: fino/src/types/SymDate.jl
# types/SymDate.jl

using Dates

"""
    SymDate

A pair of symbol and date for grouping by earnings release dates.

Fields:
- `symbol::String` — ticker symbol (stored uppercase)
- `date::Date` — the date
"""
struct SymDate
    symbol::String
    date::Date

    function SymDate(symbol::String, date::Date)
        new(uppercase(symbol), date)
    end
end

# ── Comparison and hashing ────────────────────────────────────────────────────

Base.:(==)(sd1::SymDate, sd2::SymDate) = (sd1.symbol == uppercase(sd2.symbol)) && (sd1.date == sd2.date)

    Base.hash(sd::SymDate) = hash((sd.symbol, sd.date))

# ── String representation ──────────────────────────────────────────────────────

    Base.repr(sd::SymDate) = "$(sd.symbol) - $(Dates.format(sd.date, "yyyy-mm-dd"))"

    Base.string(sd::SymDate) = repr(sd)

File: options/types/sym_date.py
from datetime import date
from dataclasses import dataclass


@dataclass
class SymDate:
    """For grouping by earnings release dates"""
    symbol: str
    date: date

    def __post_init__(self):
        self.symbol = self.symbol.upper()

    def __eq__(self, other):
        return self.symbol == other.symbol.upper() and self.date == other.date

    def __hash__(self):
        return hash((self.symbol, self.date))

    def __repr__(self):
        return f"{self.symbol} - {self.date.isoformat()}"
